utils/src/fiemap.c

```diff
@@ -0,0 +1,195 @@
+#include <unistd.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/ioctl.h>
+#include <linux/fs.h>
+#include <linux/fiemap.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <string.h>
+#include <argp.h>
+
+#include "sparse.h"
+#include "parse.h"
+#include "util.h"
+#include "format.h"
+#include "ioctl.h"
+#include "cmd.h"
+
+/*
+ * This is wholly modeled after e2fsprogs/filefrag.c from tso
+ */
+
+struct get_fiemap_args {
+	char *filename;
+	bool phys;
+	bool logical;
+	bool byte;
+};
+
+static int do_get_fiemap(struct get_fiemap_args *args)
+{
+	__u64 buf[2048];        /* __u64 for proper field alignment */
+	struct stat st;
+	struct fiemap *fiemap = (struct fiemap *)buf;
+	struct fiemap_extent *fm_ext = &fiemap->fm_extents[0];
+	int count = (sizeof(buf) - sizeof(*fiemap)) /
+			sizeof(struct fiemap_extent);
+	int fd;
+	int ret;
+	int i;
+	int nr = 0; /* XXX we could put this in fm_start to make start/count an option */
+	int last = 0;
+	int written = 0;
+	u64 off_p, off_l;
+	u64 len;
+
+	memset(fiemap, 0, sizeof(struct fiemap));
+
+	fd = open(args->filename, O_RDONLY);
+	if (fd < 0) {
+		ret = -errno;
+		fprintf(stderr, "failed to open '%s': %s (%d)\n",
+			args->filename, strerror(errno), errno);
+		goto out;
+	}
+
+	/* get block size from stat */
+	if (fstat(fd, &st) != 0) {
+		ret = -errno;
+		fprintf(stderr, "stat failed on '%s': %s (%d)\n",
+			args->filename, strerror(errno), errno);
+		goto out;
+	};
+
+	do {
+		fiemap->fm_length = ~0ULL;
+		fiemap->fm_extent_count = count;
+
+		ret = ioctl(fd, FS_IOC_FIEMAP, (unsigned long) fiemap);
+		if (ret < 0) {
+			ret = -errno;
+			fprintf(stderr, "get_fiemap ioctl failed: "
+				"%s (%d)\n", strerror(errno), errno);
+			goto out;
+		}
+
+		/* nothing returned, so exit */
+		if (fiemap->fm_mapped_extents == 0)
+			break;
+
+		for (i = 0; i < fiemap->fm_mapped_extents; i++) {
+			off_p = fm_ext[i].fe_physical;
+			off_l = fm_ext[i].fe_logical;
+			len = fm_ext[i].fe_length;
+
+			if (!args->byte) {
+				off_p /= st.st_blksize;
+				off_l /= st.st_blksize;
+				len /= st.st_blksize;
+			}
+
+			printf("%d: offset: ", nr++);
+
+			if (!args->phys)
+				printf("%llu ", off_l);
+			else if (!args->logical)
+				printf("%llu ", off_p);
+			else
+				printf("%llu %llu ", off_l, off_p);
+
+			printf("length: %llu flags: %c%c%c\n",
+				len,
+				(fm_ext[i].fe_flags & FIEMAP_EXTENT_UNKNOWN) ? 'O' : '.',
+				(fm_ext[i].fe_flags & FIEMAP_EXTENT_UNWRITTEN) ? 'U' : '.',
+				(fm_ext[i].fe_flags & FIEMAP_EXTENT_LAST) ? 'L' : '.');
+
+			if (fm_ext[i].fe_flags & FIEMAP_EXTENT_LAST)
+				last = 1;
+
+			if ((fm_ext[i].fe_flags & FIEMAP_EXTENT_UNWRITTEN) == 0)
+				written++;
+		}
+
+		/* fm_start from the next logical extent */
+		fiemap->fm_start = fm_ext[i-1].fe_logical + fm_ext[i-1].fe_length;
+	} while (last == 0);
+
+	printf("extents: %u\n", written);
```

Comment:
(as mentioned elsewhere..) I still don't understand why we're only counting written extents.  Just print nr here?  (And make it larger than an int for good measure, I guess, but not really critical.  It'd be nice to not have to think about the embarrassing overflowed count of extents.)

Reply:
We'll go for plain `nr` here to avoid any of the confusion. You'll see in the test output that there will be this change:

```
== goofy preallocation amounts work
/mnt/test/test/data-prealloc/file-1: 5 extents found
/mnt/test/test/data-prealloc/file-2: 5 extents found
/mnt/test/test/data-prealloc/file-1: 5 extents found
/mnt/test/test/data-prealloc/file-2: 5 extents found
/mnt/test/test/data-prealloc/file-1: 3 extents found
/mnt/test/test/data-prealloc/file-2: 3 extents found
/mnt/test/test/data-prealloc/file-1: extents: 6
/mnt/test/test/data-prealloc/file-2: extents: 6
/mnt/test/test/data-prealloc/file-1: extents: 6
/mnt/test/test/data-prealloc/file-2: extents: 6
/mnt/test/test/data-prealloc/file-1: extents: 3
/mnt/test/test/data-prealloc/file-2: extents: 3
== block writes into region allocs hole
```

These numbers won't match because previously an unwritten (or, unmerged?) thing ended up avoided being counted (in filefrag) but now they will be counted. The rest of the numbers didn't meet this case and will match.